Add Take Observable operator

keithamus opened the pull request:
#### 8bf6f2dcf6ac23fbdbc04fe01f6f0d60d035d35f
<pre>
Add Take Observable operator
<a href="https://bugs.webkit.org/show_bug.cgi?id=277103">https://bugs.webkit.org/show_bug.cgi?id=277103</a>

Reviewed by Darin Adler.

This adds the .take(amount) operator for Observables; implementing this
behind a InternalObserverTake backing class and subsequent
SubscribeCallback specialisation.

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-take.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-take.any.worker-expected.txt:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSSubscriberCustom.cpp:
* Source/WebCore/dom/InternalObserverFromScript.cpp:
* Source/WebCore/dom/InternalObserverTake.cpp: Added.
  Implements createSubscriberCallbackTake as well as the
  InternalObserverTake and SubscribeCallbackTake classes.
(WebCore::createSubscriberCallbackTake):
* Source/WebCore/dom/InternalObserverTake.h: Copied from Source/WebCore/dom/Observable.idl.
  Adds the createSubscriberCallbackTake function which creates a callback to
  handle the logic for `take()`.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::create):
  Drop `ExceptionOr&lt;&gt;` which was unused, and made it difficult to easily
  create new operators (which also don&apos;t need `ExceptionOr&lt;&gt;`).
(WebCore::Observable::take):
  Implement the take() operator.
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.idl:
* Source/WebCore/dom/Subscriber.cpp:
(WebCore::Subscriber::observerConcurrently):
  Moved out of the header file during refactor to add
  visitAdditionalChildren.
(WebCore::Subscriber::visitAdditionalChildren):
  Added to traverse into InternalObserver to allow for objects in
  InternalObserver to be marked.
* Source/WebCore/dom/Subscriber.h:

Canonical link: <a href="https://commits.webkit.org/281460@main">https://commits.webkit.org/281460@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3d1fe37bcd027df2f395ca178302803ddf9089c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59980 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39328 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12532 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63898 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10510 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62109 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47000 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10712 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48604 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7331 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62010 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36672 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51935 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29446 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33377 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9182 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9430 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55289 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9462 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65630 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3910 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9319 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55962 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3928 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51930 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56102 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13296 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3242 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35141 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36223 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37311 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35967 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/bindings/js/JSSubscriberCustom.cpp:
```diff
@@ -25,6 +25,7 @@
 
 #include "config.h"
 #include "JSSubscriber.h"
+#include "VoidCallback.h"
```
WebKit coding style puts a blank line between the main header and a paragraph with the other included headers. Here, the main header is `JSSubscriber.h`.

darinadler commented on Source/WebCore/dom/InternalObserverTake.cpp:
```diff
@@ -0,0 +1,104 @@
+InternalObserverTake::InternalObserverTake(ScriptExecutionContext& context, Ref<Subscriber> subscriber, uint64_t amount)
+    : InternalObserver(context)
+    , m_subscriber(subscriber)
+    , m_amount(amount) { }
```
WebKit coding style puts the `{ }` on a separate line, outdented.

darinadler commented on Source/WebCore/dom/InternalObserverTake.cpp:
```diff
@@ -0,0 +1,104 @@
+InternalObserverTake::SubscriberCallbackTake::SubscriberCallbackTake(ScriptExecutionContext& context, Ref<Observable> source, uint64_t amount)
+    : SubscriberCallback(&context)
+    , m_sourceObservable(source)
+    , m_amount(amount) { }
```
WebKit coding style puts the `{ }` on a separate line, outdented.

darinadler commented on Source/WebCore/dom/InternalObserverTake.h:
```diff
@@ -0,0 +1,71 @@
+
+#include "InternalObserver.h"
+#include "SubscriberCallback.h"
+#include <wtf/RefCounted.h>
```
Why is this included here? I don’t see any use of it in the header.

darinadler commented on Source/WebCore/dom/InternalObserverTake.h:
```diff
@@ -0,0 +1,71 @@
+
+class Observable;
+class ScriptExecutionContext;
+class Subscriber;
```
Is this needed? I’d think `SubscriberCallback.h` would forward declare this.

darinadler commented on Source/WebCore/dom/InternalObserverTake.h:
```diff
@@ -0,0 +1,71 @@
+        Ref<Observable> m_sourceObservable;
+        uint64_t m_amount;
+
+        bool hasCallback() const { return true; }
```
This should specify either `override` or `final` to be explicit about the fact that we are intending to override a virtual function here.

darinadler commented on Source/WebCore/dom/InternalObserverTake.h:
```diff
@@ -0,0 +1,71 @@
+    public:
+        static Ref<SubscriberCallbackTake> create(ScriptExecutionContext&, Ref<Observable>, uint64_t);
+
+        explicit SubscriberCallbackTake(ScriptExecutionContext&, Ref<Observable>, uint64_t);
```
This constructor should be `private` since we have a `create` function and that is the only function that should call the constructor. We also don’t usually put `explicit` on constructors with multiple arguments, in part for historical reasons, but also because the value of specifying `explicit` for such cases is minimal.

darinadler commented on Source/WebCore/dom/InternalObserverTake.h:
```diff
@@ -0,0 +1,71 @@
+
+        explicit SubscriberCallbackTake(ScriptExecutionContext&, Ref<Observable>, uint64_t);
+
+        CallbackResult<void> handleEvent(Subscriber&) final;
```
This override should be `private`, unless there is some need to call this directly.

keithamus replied:
I believe this gets called by `Observable.cpp`.

darinadler replied:
Directly or virtually? It can still be private if called via a base class.

darinadler commented on Source/WebCore/dom/InternalObserverTake.h:
```diff
@@ -0,0 +1,71 @@
+public:
+    static Ref<InternalObserverTake> create(ScriptExecutionContext&, Ref<Subscriber>, uint64_t);
+
+    explicit InternalObserverTake(ScriptExecutionContext&, Ref<Subscriber>, uint64_t);
```
This constructor should be `private` since we have a `create` function and that is the only function that should call the constructor. We also don’t usually put `explicit` on constructors with multiple arguments, in part for historical reasons, but also because the value of specifying `explicit` for such cases is minimal.

darinadler commented on Source/WebCore/dom/InternalObserverTake.h:
```diff
@@ -0,0 +1,71 @@
+    void complete() final;
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor&) const final;
+    void visitAdditionalChildren(JSC::SlotVisitor&) const final;
```
All of these overrides can be made `private` unless there is a need to call them directly.

darinadler commented on Source/WebCore/dom/InternalObserverTake.h:
```diff
@@ -0,0 +1,71 @@
+class ScriptExecutionContext;
+class Subscriber;
+
+class InternalObserverTake final : public InternalObserver {
```
We should consider leaving these classes out of the header completely and instead simply exposing a function like this:
```
Ref<Observable> createSubscriberCallbackTake(ScriptExecutionContext&, Ref<Observable>, uint64_t);
```
Nothing else would need to be in the header. The implementation details need not be exposed to the file using this class. That way any implementation detail changes would not cause any other file to recompile. The classes can be completely contained inside the `.cpp` file.

There might be an even better name for this function; it doesn’t need to name the class, but rather describe the purpose.

darinadler commented on Source/WebCore/dom/Subscriber.cpp:
```diff
@@ -148,6 +151,35 @@ void Subscriber::reportErrorObject(JSC::JSValue value)
     reportException(globalObject, JSC::Exception::create(vm, value));
 }
 
+Vector<VoidCallback*> Subscriber::teardownCallbacksConcurrently()
+{
+    Locker locker { m_teardownsLock };
+    return m_teardowns.map([](auto& callback) {
+        return callback.ptr();
+    });
+}
```
What’s the rationale for moving this out of the header? Might be a good idea, but the change log doesn’t seem to say why.

darinadler commented on Source/WebCore/dom/Subscriber.cpp:
```diff
@@ -148,6 +151,35 @@ void Subscriber::reportErrorObject(JSC::JSValue value)
+    });
+}
+
+InternalObserver* Subscriber::observerConcurrently()
+{
+    return &m_observer.get();
+}
```
What’s the rationale for moving this out of the header? Might be a good idea, but the change log doesn’t seem to say why.

darinadler commented on Source/WebCore/dom/SubscriberCallback.h:
```diff
@@ -35,7 +35,8 @@ class Subscriber;
 public:
-    using ActiveDOMCallback::ActiveDOMCallback;
+    explicit SubscriberCallback(ScriptExecutionContext* context)
+        : ActiveDOMCallback(context) { }
```
What’s the reason behind this change?

darinadler commented on Source/WebCore/dom/InternalObserverTake.cpp:
```diff
@@ -0,0 +1,104 @@
+void InternalObserverTake::next(JSC::JSValue value)
+{
+    m_subscriber->next(value);
+    m_amount -= 1;
```
Is there something that guarantees this cannot be called if `m_amount` is already `0`? Otherwise we could underflow here.

keithamus replied:
Yes `next()` won't be called after `complete()`. I'll add a guard though, just in case.

darinadler commented on Source/WebCore/dom/InternalObserverTake.h:
```diff
@@ -0,0 +1,36 @@
+
+#pragma once
+
+namespace WebCore {
```
This file technically needs an include of `<wtf/Forward.h>`.

darinadler commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -93,6 +94,12 @@ void Observable::subscribeInternal(ScriptExecutionContext& context, Ref<Internal
     }
 }
 
+
```
Extra blank line here I would omit.

darinadler commented on Source/WebCore/dom/Observable.idl:
```diff
@@ -32,4 +32,6 @@ typedef (SubscriptionObserverCallback or SubscriptionObserver) ObserverUnion;
 interface Observable {
   [RaisesException] constructor(SubscriberCallback callback);
```
I do not think this needs `[RaisesException]` after we changed the `create` function to never return an exception.

keithamus committed:
Add Take Observable operator
https://bugs.webkit.org/show_bug.cgi?id=277103

Reviewed by Darin Adler.

This adds the .take(amount) operator for Observables; implementing this
behind a InternalObserverTake backing class and subsequent
SubscribeCallback specialisation.

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-take.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-take.any.worker-expected.txt:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSSubscriberCustom.cpp:
* Source/WebCore/dom/InternalObserverFromScript.cpp:
* Source/WebCore/dom/InternalObserverTake.cpp: Added.
  Implements createSubscriberCallbackTake as well as the
  InternalObserverTake and SubscribeCallbackTake classes.
(WebCore::createSubscriberCallbackTake):
* Source/WebCore/dom/InternalObserverTake.h: Copied from Source/WebCore/dom/Observable.idl.
  Adds the createSubscriberCallbackTake function which creates a callback to
  handle the logic for `take()`.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::create):
  Drop `ExceptionOr<>` which was unused, and made it difficult to easily
  create new operators (which also don't need `ExceptionOr<>`).
(WebCore::Observable::take):
  Implement the take() operator.
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.idl:
* Source/WebCore/dom/Subscriber.cpp:
(WebCore::Subscriber::observerConcurrently):
  Moved out of the header file during refactor to add
  visitAdditionalChildren.
(WebCore::Subscriber::visitAdditionalChildren):
  Added to traverse into InternalObserver to allow for objects in
  InternalObserver to be marked.
* Source/WebCore/dom/Subscriber.h:

Canonical link: https://commits.webkit.org/281460@main
